File: Tests/RedlineCoreTests/StoreTests.swift
// Exercises the on-disk scanners against synthetic transcripts in a temp dir.
import XCTest
@testable import RedlineCore

private func makeTempDir() throws -> URL {
    let dir = URL(fileURLWithPath: NSTemporaryDirectory())
        .appendingPathComponent("redline-tests-" + UUID().uuidString)
    try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
    return dir
}

final class CodexStoreTests: XCTestCase {
    private var dir: URL!

    override func setUpWithError() throws {
        dir = try makeTempDir()
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private func write(_ lines: [String], to name: String = "rollout.jsonl") throws {
        try lines.joined(separator: "\n")
            .write(to: dir.appendingPathComponent(name), atomically: true, encoding: .utf8)
    }

    // resetsAt defaults far into the future so windows survive the unexpired() filter;
    // tests that care about rollover pass a past value explicitly.
    private func tokenCount(ts: String, pct: Double, lastInput: Int, lastCached: Int,
                            lastOutput: Int, reasoning: Int = 0,
                            resetsAt: Double = 4_000_000_000) -> String {
        let payload: [String: Any] = [
            "type": "token_count",
            "rate_limits": [
                "primary": ["used_percent": pct, "window_minutes": 300,
                            "resets_at": resetsAt],
            ],
            "info": [
                "last_token_usage": [
                    "input_tokens": lastInput,
                    "cached_input_tokens": lastCached,
                    "output_tokens": lastOutput,
                    "reasoning_output_tokens": reasoning,
                ],
            ],
        ]
        let obj: [String: Any] = ["type": "event_msg", "timestamp": ts, "payload": payload]
        let data = try! JSONSerialization.data(withJSONObject: obj)
        return String(data: data, encoding: .utf8)!
    }

    func testParsesTokensAndSplitsCachedOutOfInput() throws {
        let now = Date()
        let ts = ISO8601DateFormatter().string(from: now)
        try write([tokenCount(ts: ts, pct: 3, lastInput: 1000, lastCached: 900,
                              lastOutput: 50, reasoning: 5)])
        let snap = CodexStore(root: dir).scan(lookbackDays: 7, now: now)
        XCTAssertEqual(snap.entries.count, 1)
        let e = snap.entries[0]
        XCTAssertEqual(e.provider, "Codex")
        XCTAssertEqual(e.input, 100, "cached tokens must not be billed as fresh input")
        XCTAssertEqual(e.cacheRead, 900)
        XCTAssertEqual(e.output, 55, "reasoning tokens count as output")
    }

    func testSumsPerTurnDeltasRatherThanCumulativeTotals() throws {
        let now = Date()
        let f = ISO8601DateFormatter()
        try write([
            tokenCount(ts: f.string(from: now.addingTimeInterval(-60)),
                       pct: 1, lastInput: 100, lastCached: 0, lastOutput: 10),
            tokenCount(ts: f.string(from: now),
                       pct: 2, lastInput: 200, lastCached: 0, lastOutput: 20),
        ])
        let snap = CodexStore(root: dir).scan(lookbackDays: 7, now: now)
        let io = snap.entries.reduce(0) { $0 + $1.input + $1.output }
        XCTAssertEqual(io, 330, "each event is a delta, so they add rather than replace")
    }

    func testLimitsComeFromNewestEvent() throws {
        let now = Date()
        let f = ISO8601DateFormatter()
        try write([
            tokenCount(ts: f.string(from: now.addingTimeInterval(-600)),
                       pct: 11, lastInput: 1, lastCached: 0, lastOutput: 1),
            tokenCount(ts: f.string(from: now),
                       pct: 77, lastInput: 1, lastCached: 0, lastOutput: 1),
        ])
        let snap = CodexStore(root: dir).scan(lookbackDays: 7, now: now)
        XCTAssertEqual(snap.limits.first?.utilization, 77)
    }

    func testDropsLimitWindowsThatAlreadyReset() throws {
        // A 5-hour window that reset long ago must not be shown as current, which is exactly
        // what reading days-old Codex sessions off disk would otherwise do.
        let now = Date()
        let ts = ISO8601DateFormatter().string(from: now)
        try write([tokenCount(ts: ts, pct: 88, lastInput: 10, lastCached: 0, lastOutput: 1,
                              resetsAt: now.timeIntervalSince1970 - 3600)])
        let snap = CodexStore(root: dir).scan(lookbackDays: 7, now: now)
        XCTAssertTrue(snap.limits.isEmpty, "an expired window is meaningless, not just stale")
        XCTAssertEqual(snap.entries.count, 1, "token totals stay valid regardless")
    }

    func testIgnoresEventsOutsideLookback() throws {
        let now = Date()
        let old = ISO8601DateFormatter().string(from: now.addingTimeInterval(-40 * 86400))
        try write([tokenCount(ts: old, pct: 5, lastInput: 10, lastCached: 0, lastOutput: 10)])
        XCTAssertTrue(CodexStore(root: dir).scan(lookbackDays: 7, now: now).entries.isEmpty)
    }

    func testMissingDirectoryIsNotAnError() {
        let missing = dir.appendingPathComponent("nope")
        let snap = CodexStore(root: missing).scan(lookbackDays: 7)
        XCTAssertTrue(snap.entries.isEmpty)
        XCTAssertTrue(snap.limits.isEmpty)
    }

    func testSkipsMalformedLines() throws {
        let ts = ISO8601DateFormatter().string(from: Date())
        try write(["{not json", "", tokenCount(ts: ts, pct: 1, lastInput: 10,
                                               lastCached: 0, lastOutput: 1)])
        XCTAssertEqual(CodexStore(root: dir).scan(lookbackDays: 7).entries.count, 1)
    }
}

final class OllamaStoreTests: XCTestCase {
    private var dir: URL!
    private var log: URL!

    override func setUpWithError() throws {
        dir = try makeTempDir()
        log = dir.appendingPathComponent("ollama.jsonl")
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    func testParsesEvalCounts() throws {
        let ts = ISO8601DateFormatter().string(from: Date())
        let line = #"{"ts":"\#(ts)","model":"qwen3-coder:30b","prompt_eval_count":1200,"eval_count":340}"#
        try line.write(to: log, atomically: true, encoding: .utf8)
        let out = OllamaStore(log: log).scan(lookbackDays: 7)
        XCTAssertEqual(out.count, 1)
        XCTAssertEqual(out[0].provider, "Ollama")
        XCTAssertEqual(out[0].input, 1200)
        XCTAssertEqual(out[0].output, 340)
        XCTAssertEqual(out[0].model, "qwen3-coder:30b")
    }

    func testLocalUsageStaysUnpricedSoItIsNotCountedAsSpend() throws {
        let ts = ISO8601DateFormatter().string(from: Date())
        let line = #"{"ts":"\#(ts)","model":"qwen3-coder:30b","prompt_eval_count":10,"eval_count":5}"#
        try line.write(to: log, atomically: true, encoding: .utf8)
        let entries = OllamaStore(log: log).scan(lookbackDays: 7)
        let a = aggregate(entries, since: Date().addingTimeInterval(-3600), config: Config())
        XCTAssertEqual(a.io, 15)
        XCTAssertEqual(a.cost, 0, "local inference has no dollar cost")
        XCTAssertTrue(a.hasUnpriced)
    }

    func testMissingLogIsNotAnError() {
        let store = OllamaStore(log: dir.appendingPathComponent("absent.jsonl"))
        XCTAssertFalse(store.isConfigured)
        XCTAssertTrue(store.scan(lookbackDays: 7).isEmpty)
    }

    func testZeroTokenCallsAreSkipped() throws {
        let ts = ISO8601DateFormatter().string(from: Date())
        let line = #"{"ts":"\#(ts)","model":"m","prompt_eval_count":0,"eval_count":0}"#
        try line.write(to: log, atomically: true, encoding: .utf8)
        XCTAssertTrue(OllamaStore(log: log).scan(lookbackDays: 7).isEmpty)
    }
}

final class ClaudeStoreTests: XCTestCase {
    private var dir: URL!

    override func setUpWithError() throws {
        dir = try makeTempDir()
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: dir)
    }

    private func line(id: String, requestId: String, ts: String, model: String,
                      input: Int, output: Int, cacheRead: Int = 0) -> String {
        let obj: [String: Any] = [
            "timestamp": ts,
            "requestId": requestId,
            "message": [
                "id": id,
                "model": model,
                "usage": [
                    "input_tokens": input,
                    "output_tokens": output,
                    "cache_read_input_tokens": cacheRead,
                ],
            ],
        ]
        return String(data: try! JSONSerialization.data(withJSONObject: obj), encoding: .utf8)!
    }

    func testDedupesRepeatedMessageIdsAcrossFiles() throws {
        let now = Date()
        let ts = ISO8601DateFormatter().string(from: now)
        let l = line(id: "msg_1", requestId: "req_1", ts: ts,
                     model: "claude-sonnet-5", input: 100, output: 10)
        // A resumed session copies identical ids into a second transcript
        for name in ["a.jsonl", "b.jsonl"] {
            try l.write(to: dir.appendingPathComponent(name), atomically: true, encoding: .utf8)
        }
        let out = UsageStore(root: dir).scan(lookbackDays: 7, now: now)
        XCTAssertEqual(out.count, 1, "the same message must not be counted twice")
    }

    func testWideningTheRangeRereadsCachedFiles() throws {
        let now = Date()
        let iso = ISO8601DateFormatter()
        let body = [line(id: "m1", requestId: "r1",
                         ts: iso.string(from: now.addingTimeInterval(-3 * 86400)),
                         model: "claude-sonnet-5", input: 10, output: 1),
                    line(id: "m2", requestId: "r2",
                         ts: iso.string(from: now.addingTimeInterval(-20 * 86400)),
                         model: "claude-sonnet-5", input: 20, output: 2)]
            .joined(separator: "\n")
        try body.write(to: dir.appendingPathComponent("a.jsonl"), atomically: true,
                       encoding: .utf8)
        // One store across both scans, which is what the dashboard does on a range change
        let store = UsageStore(root: dir)
        XCTAssertEqual(store.scan(lookbackDays: 7, now: now).count, 1)
        XCTAssertEqual(store.scan(lookbackDays: 30, now: now).count, 2,
                       "widening the range must re-read files cached for a narrower one")
        XCTAssertEqual(store.scan(lookbackDays: 7, now: now).count, 1,
                       "narrowing again must not leak the wider window's entries")
    }

    func testSkipsSyntheticModel() throws {
        let ts = ISO8601DateFormatter().string(from: Date())
        try line(id: "m", requestId: "r", ts: ts, model: "<synthetic>", input: 5, output: 5)
            .write(to: dir.appendingPathComponent("s.jsonl"), atomically: true, encoding: .utf8)
        XCTAssertTrue(UsageStore(root: dir).scan(lookbackDays: 7).isEmpty)
    }

    func testMissingDirectoryIsNotAnError() {
        XCTAssertTrue(UsageStore(root: dir.appendingPathComponent("nope"))
            .scan(lookbackDays: 7).isEmpty)
    }
}

File: windows/tests/Redline.Core.Tests/StoreTests.cs
// Exercises the on-disk scanners against synthetic transcripts in a temp dir.
using System.Globalization;
using System.Text.Json.Nodes;
using Redline.Core;

namespace Redline.Core.Tests;

internal static class StoreFixtures
{
    public static string MakeTempDir()
    {
        var dir = Path.Combine(Path.GetTempPath(), "redline-tests-" + Guid.NewGuid());
        Directory.CreateDirectory(dir);
        return dir;
    }

    public static void Remove(string dir)
    {
        try { Directory.Delete(dir, recursive: true); } catch { }
    }

    /// What ISO8601DateFormatter().string(from:) writes: whole seconds, UTC, trailing Z.
    public static string Iso(DateTimeOffset d) =>
        d.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);
}

public sealed class CodexStoreTests : IDisposable
{
    private readonly string _dir = StoreFixtures.MakeTempDir();

    public void Dispose() => StoreFixtures.Remove(_dir);

    private void Write(string[] lines, string name = "rollout.jsonl") =>
        File.WriteAllText(Path.Combine(_dir, name), string.Join("\n", lines));

    // resetsAt defaults far into the future so windows survive Unexpired();
    // tests that care about rollover pass a past value explicitly.
    private static string TokenCount(string ts, double pct, int lastInput, int lastCached,
                                     int lastOutput, int reasoning = 0, double resetsAt = 4_000_000_000)
    {
        var payload = new JsonObject
        {
            ["type"] = "token_count",
            ["rate_limits"] = new JsonObject
            {
                ["primary"] = new JsonObject { ["used_percent"] = pct, ["window_minutes"] = 300, ["resets_at"] = resetsAt },
            },
            ["info"] = new JsonObject
            {
                ["last_token_usage"] = new JsonObject
                {
                    ["input_tokens"] = lastInput,
                    ["cached_input_tokens"] = lastCached,
                    ["output_tokens"] = lastOutput,
                    ["reasoning_output_tokens"] = reasoning,
                },
            },
        };
        return new JsonObject { ["type"] = "event_msg", ["timestamp"] = ts, ["payload"] = payload }.ToJsonString();
    }

    [Fact]
    public void ParsesTokensAndSplitsCachedOutOfInput()
    {
        var now = DateTimeOffset.UtcNow;
        Write(new[] { TokenCount(StoreFixtures.Iso(now), 3, 1000, 900, 50, reasoning: 5) });
        var snap = new CodexStore(_dir).Scan(7, now);
        Assert.Single(snap.Entries);
        var e = snap.Entries[0];
        Assert.Equal("Codex", e.Provider);
        Assert.Equal(100, e.Input); // cached tokens must not be billed as fresh input
        Assert.Equal(900, e.CacheRead);
        Assert.Equal(55, e.Output); // reasoning tokens count as output
    }

    [Fact]
    public void SumsPerTurnDeltasRatherThanCumulativeTotals()
    {
        var now = DateTimeOffset.UtcNow;
        Write(new[]
        {
            TokenCount(StoreFixtures.Iso(now.AddSeconds(-60)), 1, 100, 0, 10),
            TokenCount(StoreFixtures.Iso(now), 2, 200, 0, 20),
        });
        var snap = new CodexStore(_dir).Scan(7, now);
        var io = snap.Entries.Sum(e => e.Input + e.Output);
        Assert.Equal(330, io); // each event is a delta, so they add rather than replace
    }

    [Fact]
    public void LimitsComeFromNewestEvent()
    {
        var now = DateTimeOffset.UtcNow;
        Write(new[]
        {
            TokenCount(StoreFixtures.Iso(now.AddSeconds(-600)), 11, 1, 0, 1),
            TokenCount(StoreFixtures.Iso(now), 77, 1, 0, 1),
        });
        var snap = new CodexStore(_dir).Scan(7, now);
        Assert.Equal(77.0, snap.Limits.FirstOrDefault()?.Utilization);
    }

    [Fact]
    public void DropsLimitWindowsThatAlreadyReset()
    {
        // A 5-hour window that reset long ago must not be shown as current, which is exactly
        // what reading days-old Codex sessions off disk would otherwise do.
        var now = DateTimeOffset.UtcNow;
        Write(new[] { TokenCount(StoreFixtures.Iso(now), 88, 10, 0, 1,
                                 resetsAt: now.ToUnixTimeMilliseconds() / 1000.0 - 3600) });
        var snap = new CodexStore(_dir).Scan(7, now);
        Assert.Empty(snap.Limits); // an expired window is meaningless, not just stale
        Assert.Single(snap.Entries); // token totals stay valid regardless
    }

    [Fact]
    public void IgnoresEventsOutsideLookback()
    {
        var now = DateTimeOffset.UtcNow;
        Write(new[] { TokenCount(StoreFixtures.Iso(now.AddDays(-40)), 5, 10, 0, 10) });
        Assert.Empty(new CodexStore(_dir).Scan(7, now).Entries);
    }

    [Fact]
    public void MissingDirectoryIsNotAnError()
    {
        var snap = new CodexStore(Path.Combine(_dir, "nope")).Scan(7);
        Assert.Empty(snap.Entries);
        Assert.Empty(snap.Limits);
    }

    [Fact]
    public void SkipsMalformedLines()
    {
        var ts = StoreFixtures.Iso(DateTimeOffset.UtcNow);
        Write(new[] { "{not json", "", TokenCount(ts, 1, 10, 0, 1) });
        Assert.Single(new CodexStore(_dir).Scan(7).Entries);
    }
}

public sealed class OllamaStoreTests : IDisposable
{
    private readonly string _dir = StoreFixtures.MakeTempDir();
    private string Log => Path.Combine(_dir, "ollama.jsonl");

    public void Dispose() => StoreFixtures.Remove(_dir);

    [Fact]
    public void ParsesEvalCounts()
    {
        var ts = StoreFixtures.Iso(DateTimeOffset.UtcNow);
        File.WriteAllText(Log, $"{{\"ts\":\"{ts}\",\"model\":\"qwen3-coder:30b\",\"prompt_eval_count\":1200,\"eval_count\":340}}");
        var output = new OllamaStore(Log).Scan(7);
        Assert.Single(output);
        Assert.Equal("Ollama", output[0].Provider);
        Assert.Equal(1200, output[0].Input);
        Assert.Equal(340, output[0].Output);
        Assert.Equal("qwen3-coder:30b", output[0].Model);
    }

    [Fact]
    public void LocalUsageStaysUnpricedSoItIsNotCountedAsSpend()
    {
        var ts = StoreFixtures.Iso(DateTimeOffset.UtcNow);
        File.WriteAllText(Log, $"{{\"ts\":\"{ts}\",\"model\":\"qwen3-coder:30b\",\"prompt_eval_count\":10,\"eval_count\":5}}");
        var entries = new OllamaStore(Log).Scan(7);
        var a = Usage.Aggregate(entries, DateTimeOffset.UtcNow.AddSeconds(-3600), new Config());
        Assert.Equal(15, a.Io);
        Assert.Equal(0.0, a.Cost); // local inference has no dollar cost
        Assert.True(a.HasUnpriced);
    }

    [Fact]
    public void MissingLogIsNotAnError()
    {
        var store = new OllamaStore(Path.Combine(_dir, "absent.jsonl"));
        Assert.False(store.IsConfigured);
        Assert.Empty(store.Scan(7));
    }

    [Fact]
    public void ZeroTokenCallsAreSkipped()
    {
        var ts = StoreFixtures.Iso(DateTimeOffset.UtcNow);
        File.WriteAllText(Log, $"{{\"ts\":\"{ts}\",\"model\":\"m\",\"prompt_eval_count\":0,\"eval_count\":0}}");
        Assert.Empty(new OllamaStore(Log).Scan(7));
    }
}

public sealed class ClaudeStoreTests : IDisposable
{
    private readonly string _dir = StoreFixtures.MakeTempDir();

    public void Dispose() => StoreFixtures.Remove(_dir);

    private static string Line(string id, string requestId, string ts, string model,
                               int input, int output, int cacheRead = 0) =>
        new JsonObject
        {
            ["timestamp"] = ts,
            ["requestId"] = requestId,
            ["message"] = new JsonObject
            {
                ["id"] = id,
                ["model"] = model,
                ["usage"] = new JsonObject
                {
                    ["input_tokens"] = input,
                    ["output_tokens"] = output,
                    ["cache_read_input_tokens"] = cacheRead,
                },
            },
        }.ToJsonString();

    [Fact]
    public void DedupesRepeatedMessageIdsAcrossFiles()
    {
        var now = DateTimeOffset.UtcNow;
        var l = Line("msg_1", "req_1", StoreFixtures.Iso(now), "claude-sonnet-5", 100, 10);
        // A resumed session copies identical ids into a second transcript
        foreach (var name in new[] { "a.jsonl", "b.jsonl" }) File.WriteAllText(Path.Combine(_dir, name), l);
        var output = new UsageStore(_dir).Scan(7, now);
        Assert.Single(output); // the same message must not be counted twice
    }

    [Fact]
    public void WideningTheRangeRereadsCachedFiles()
    {
        var now = DateTimeOffset.UtcNow;
        var body = string.Join("\n",
            Line("m1", "r1", StoreFixtures.Iso(now.AddDays(-3)), "claude-sonnet-5", 10, 1),
            Line("m2", "r2", StoreFixtures.Iso(now.AddDays(-20)), "claude-sonnet-5", 20, 2));
        File.WriteAllText(Path.Combine(_dir, "a.jsonl"), body);
        // One store across both scans, which is what the dashboard does on a range change
        var store = new UsageStore(_dir);
        Assert.Single(store.Scan(7, now));
        // widening the range must re-read files cached for a narrower one
        Assert.Equal(2, store.Scan(30, now).Count);
        // narrowing again must not leak the wider window's entries
        Assert.Single(store.Scan(7, now));
    }

    [Fact]
    public void SkipsSyntheticModel()
    {
        var ts = StoreFixtures.Iso(DateTimeOffset.UtcNow);
        File.WriteAllText(Path.Combine(_dir, "s.jsonl"), Line("m", "r", ts, "<synthetic>", 5, 5));
        Assert.Empty(new UsageStore(_dir).Scan(7));
    }

    [Fact]
    public void MissingDirectoryIsNotAnError()
    {
        Assert.Empty(new UsageStore(Path.Combine(_dir, "nope")).Scan(7));
    }
}
